### PR TITLE
common,crimson: supporting admin-socket commands

### PR DESCRIPTION
**Pull request by ronen-fr**
NOTE: for review. Has not been rebased yet to latest 'master'. ...But was heavily tested.

- basic infrastructure:
  - the seastar::thread that handles commands arriving over the UNIX_domain socket
  - command dispatching
- the implementation of some commands.

Test support:
- some inline unit-testing controlled by an #ifdef
- a python script to send sequences of commands. See:https://drive.google.com/file/d/1rOrl1B5tOKWBH6kWMEHgA_cYgzjiCC8G/view?usp=sharing
https://drive.google.com/file/d/1E819x-ejEypyDSnzqfc0GRBAXyQKr0sC/view?usp=sharing
https://drive.google.com/file/d/1-MQKnRcfNh3iciOgXKhWCtGE8UNgStKK/view?usp=sharing
https://drive.google.com/file/d/19c-Xi2TyBa4e7NinKjj4BgZ712nZCykD/view?usp=sharing

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/common/admin_socket.cc [hunk 1]
```diff
@@ -401,7 +401,7 @@ int AdminSocket::execute_command(const std::string& cmd, ceph::bufferlist& out)
   }
   string args;
   if (match != cmd) {
-    args = cmd.substr(match.length() + 1);
+    args = cmd.substr(match.length() + 1); /// RRR don't understand this line
```
Comment (athanatos) [inline src/common/admin_socket.cc]: Looks to me like it's taking the portions of the matched command past the end of the matched prefix and interpreting it as arguments.

File: src/crimson/admin/admin_socket.cc [hunk 2]
```diff
@@ -0,0 +1,896 @@
...
+  it.m_in_gate = true;
+  return it;
+}
+
+AdminHooksIter AdminSocket::end() {
+  AdminHooksIter it{*this, true};
+  return it;
+}
+
+/*!
```
Comment (athanatos) [inline src/crimson/admin/admin_socket.cc]: Kind of a nit, but this doesn't really match any of the comment conventions used anywhere else in the code base (except apparently for crush.h and a couple of files in src/erasure-code).  Despite what CodingStyle says, our function/class comments tend to do the

/**
 *
 */

thing with occasionally with javadoc/kernel flavored tags for args/return/etc if that sort of thing seems useful.

File: src/common/ceph_context.h [hunk 3]
```diff
@@ -65,7 +67,7 @@ class CephContext {
 	      int = 0)
     : CephContext{}
   {}
-  CephContext(CephContext&&) = default;
+  CephContext(CephContext&&); // = default;
```
Comment (athanatos) [inline src/common/ceph_context.h]: If you're changing the constructor, go ahead and just remove the = default part.

File: src/common/ceph_context.h [hunk 4]
```diff
@@ -41,7 +43,7 @@
 
 #include "crush/CrushLocation.h"
 
-class AdminSocket;
+//class AdminSocket;  // not enough for lw_shared_ptr
```
Comment (athanatos) [inline src/common/ceph_context.h]: Same here.

File: src/crimson/admin/admin_socket.cc [hunk 5]
```diff
@@ -0,0 +1,896 @@
...
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include "common/version.h"
+#include "crimson/net/Socket.h"
+// activate for some inline unit-testing: #define UNIT_TEST_OSD_ASOK
```
Comment (athanatos) [inline src/crimson/admin/admin_socket.cc]: For merge, it would be nice to create an actual unit test to run in the usual make check paths.

File: src/crimson/admin/osd_admin.cc [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    seastar::future<> exec_command(Formatter* f, std::string_view command, const cmdmap_t& cmdmap,
+	                      std::string_view format, bufferlist& out) const final {
+
+      f->dump_stream("cluster_fsid") << m_osd_admin.osd_superblock().cluster_fsid;
+      f->dump_stream("osd_fsid") << m_osd_admin.osd_superblock().osd_fsid;
+      f->dump_unsigned("whoami", m_osd_admin.osd_superblock().whoami);
+      // \todo f->dump_string("state", get_state_name(get_state()));       // RRR where should I look for the data?
+      f->dump_unsigned("oldest_map", m_osd_admin.osd_superblock().oldest_map);
+      f->dump_unsigned("newest_map", m_osd_admin.osd_superblock().newest_map);
+      // \todo f->dump_unsigned("num_pgs", num_pgs);  -> where to find the data?
```
Comment (athanatos) [inline src/crimson/admin/osd_admin.cc]: OSD::pgs

File: src/crimson/admin/osd_admin.cc [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+  class OsdStatusHook : public OsdAdminHookBase {
+  public:
+    explicit OsdStatusHook(OsdAdminImp& master) : OsdAdminHookBase(master) {};
+    seastar::future<> exec_command(Formatter* f, std::string_view command, const cmdmap_t& cmdmap,
+	                      std::string_view format, bufferlist& out) const final {
+
+      f->dump_stream("cluster_fsid") << m_osd_admin.osd_superblock().cluster_fsid;
+      f->dump_stream("osd_fsid") << m_osd_admin.osd_superblock().osd_fsid;
+      f->dump_unsigned("whoami", m_osd_admin.osd_superblock().whoami);
+      // \todo f->dump_string("state", get_state_name(get_state()));       // RRR where should I look for the data?
```
Comment (athanatos) [inline src/crimson/admin/osd_admin.cc]: OSD::state

File: src/crimson/common/config_proxy.h [hunk 8]
```diff
@@ -28,10 +32,10 @@ class ConfigProxy : public seastar::peering_sharded_service<ConfigProxy>
   ObserverMgr<ConfigObserver> obs_mgr;
 
   const md_config_t& get_config() const {
-    return remote_config ? *remote_config : * local_config;
```
Comment (athanatos) [inline src/crimson/common/config_proxy.h]: Probably ought to be in its own commit since it's just a cosmetic change.

File: src/osd/OSD.cc [hunk 9]
```diff
@@ -285,9 +285,9 @@ OSDService::~OSDService()
 {
   delete objecter;
 
-  for (auto f : objecter_finishers) {
+  for (auto& f : objecter_finishers) {
```
Comment (athanatos) [inline src/osd/OSD.cc]: Separate commit.

File: src/crimson/admin/admin_socket.h [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+  seastar::future<> handle_client(seastar::input_stream<char>& inp, seastar::output_stream<char>& out);
+
+  seastar::future<> execute_line(std::string cmdline, seastar::output_stream<char>& out);
+
+  bool validate_command(const parsed_command_t& parsed,
+                        const std::string& command_text,
+                        ceph::buffer::list& out) const;
+
+  CephContext* m_cct;
+  bool do_die{false};  // RRR when is the OSD expected to kill the asok interface?
```
Comment (athanatos) [inline src/crimson/admin/admin_socket.h]: During shutdown, probably ought to be destroyed after things unregister.

File: src/crimson/admin/admin_socket.cc [hunk 11]
```diff
@@ -0,0 +1,896 @@
...
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2011 New Dream Network
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
```
Comment (athanatos) [inline src/crimson/admin/admin_socket.cc]: The license bit here isn't needed (though the vim and emacs mode lines are widely appreciated).

File: src/common/admin_socket.cc (hunk 1 above)
Comment (ronen-fr) [reply src/common/admin_socket.cc]: Thanks. I know that's the intention, but the match.length(), I think, does not count the
'{prefix,' that is still in cmd

File: src/crimson/admin/admin_socket.cc (hunk 2 above)
Comment (ronen-fr) [reply src/crimson/admin/admin_socket.cc]: well - that's what seastar, SPDK, and some more libraries in our code use.
Doxygen supports all three options ( ///, /**, /*! ), and I followed what I use in Seastar changes.

File: src/common/ceph_context.h (hunk 3 above)
Comment (ronen-fr) [reply src/common/ceph_context.h]: fixed

File: src/common/ceph_context.h (hunk 4 above)
Comment (ronen-fr) [reply src/common/ceph_context.h]: fixed

File: src/crimson/admin/admin_socket.cc (hunk 5 above)
Comment (ronen-fr) [reply src/crimson/admin/admin_socket.cc]: will do.

File: src/crimson/common/config_proxy.h (hunk 8 above)
Comment (ronen-fr) [reply src/crimson/common/config_proxy.h]: A general question: I'd prefer doing those minor changes as part of a larger commit, as otherwise
the effort is too large.

File: src/osd/OSD.cc (hunk 9 above)
Comment (ronen-fr) [reply src/osd/OSD.cc]: OK

File: src/crimson/admin/admin_socket.cc (hunk 11 above)
Comment (ronen-fr) [reply src/crimson/admin/admin_socket.cc]: What license header, if any, should we use in new files?
(and it's even more problematic when creating a new file from contents factored out
from an existing file with its own old copyright message).

File: src/common/admin_socket.cc (hunk 1 above)
Comment (ronen-fr) [reply src/common/admin_socket.cc]: Doesn't matter. Someone has already fixed that code.

File: src/crimson/common/config_proxy.h (hunk 8 above)
Comment (tchaikov) [reply src/crimson/common/config_proxy.h]: FWIW, maybe you could use `git add -p` to make this process simpler?

File: src/crimson/admin/admin_socket.cc (hunk 2 above)
Comment (tchaikov) [reply src/crimson/admin/admin_socket.cc]: i think we are not obliged to follow any of the coding styles of seastar, SPDK or 3rd party libraries . let me put this way, what if they don't use the same coding styles? which one shall we follow?

File: src/crimson/admin/admin_socket.cc (hunk 11 above)
Comment (tchaikov) [reply src/crimson/admin/admin_socket.cc]: see https://github.com/ceph/ceph/blob/719bff9f35b9ccf272d754058a497030df8c2366/CodingStyle#L73-L74

File: src/crimson/admin/admin_socket.cc (hunk 2 above)
Comment (athanatos) [reply src/crimson/admin/admin_socket.cc]: Yeah, let's stick with the
/**
 *
 */
style for uniformity.

File: src/crimson/common/config_proxy.h (hunk 8 above)
Comment (athanatos) [reply src/crimson/common/config_proxy.h]: I find that using git gui makes separating out little patches like that to be very quick and it makes future bug investigations a lot easier, but whatever you're comfortable with I guess.

File: src/common/ceph_context.h [hunk 12]
```diff
@@ -79,11 +81,40 @@ class CephContext {
...
+
+  /**
+   * Get the admin socket associated with this CephContext.
+   *
+   * Note that an admin socket should only be created for
+   * daemons.
+   *
+   * @return the admin socket or nullptr
+   */
+  AdminSocket* get_admin_socket() {
```
Comment (athanatos) [inline src/common/ceph_context.h]: In a vacuum, I'd assume that with this signature get_admin_socket() gives me a pointer to an AdminSocket that I own and am responsible for destructing.  If the intention is for the consumer not to take ownership, this should probably be a std::optional<AdminSocket&> or it should assert that asok exists.

File: src/crimson/admin/admin_socket.cc [hunk 13]
```diff
@@ -0,0 +1,896 @@
...
+#include "crimson/net/Socket.h"
+// activate for some inline unit-testing: #define UNIT_TEST_OSD_ASOK
+#include "crimson/admin/admin_socket.h"
+#include "crimson/common/log.h"
+//#include "seastar/testing/test_case.hh"
+#include "seastar/net/api.hh"
+#include "seastar/net/inet_address.hh"
+#include "seastar/core/reactor.hh"
+#include "seastar/core/thread.hh"
+//#include "seastar/util/log.hh"
```
Comment (athanatos) [inline src/crimson/admin/admin_socket.cc]: We can probably remove the commented out includes.

File: src/crimson/admin/admin_socket.cc [hunk 14]
```diff
@@ -0,0 +1,896 @@
...
+#include "seastar/util/std-compat.hh"
+#include <boost/algorithm/string.hpp>
+
+/*!
+  A Crimson-wise version of the admin socket - implementation file
+
+  \todo handle the unlinking of the admin socket. Note that 'cleanup_files' at-exit functionality is not yet
```
Comment (athanatos) [inline src/crimson/admin/admin_socket.cc]: CodingStyle > https://google.github.io/styleguide/cppguide.html#Line_Length

'Each line of text in your code should be at most 80 characters long.'

File: src/crimson/admin/admin_socket.cc [hunk 15]
```diff
@@ -0,0 +1,896 @@
...
+/*!
+  Hooks table - iterator support
+
+  Note: Each server-block is locked (via its gate member) before
+        iterating over its entries.
+ */
+
+AdminHooksIter AdminSocket::begin() {
+  AdminHooksIter it{*this};
+  it.m_miter->second.m_gate.enter();
```
Comment (athanatos) [inline src/crimson/admin/admin_socket.cc]: I guess the idea here is that a user of AdminHooksIter might defer to a future while holding a reference to the iterator and you want to avoid the hooks mutating during that time.  I'd suggest that we don't really need an atomic view of the underlying container, would it be possible to modify this so that "concurrent" modifications are tolerated without a lock?

Another option would be to iterate over all of the hooks synchronously and then wait on the result of all of the futures at the end -- that would eliminate the possibility of concurrent modification.

Keep in mind, even once we go to multiple reactors, each reactor will have its own set of hooks, so we're not worried about true concurrent mutation.

**Comment by athanatos** [comment]
At a high level, I think we can simply the design here.  If I'm understanding the logic here correctly, socket hook registration is centralized across reactors (with a single reactor, between the reactor where the osd is running and the thread controlling the socket itself).

Instead, what if each reactor maintains an independent (and therefore lock-free) server registration.  The admin socket thread then is only responsible for reading the input from the socket, sending it to each reactor and then waiting for and merging the responses?

File: src/crimson/admin/admin_socket.cc (hunk 2 above)
Comment (ronen-fr) [reply src/crimson/admin/admin_socket.cc]: no problem

File: src/common/ceph_context.h (hunk 12 above)
Comment (ronen-fr) [reply src/common/ceph_context.h]: OK. I agree with your reasoning in this case. Will change.
Note though that generally I only consider returning a unique_ptr as a strong hint of passing
ownership.

File: src/crimson/admin/admin_socket.cc (hunk 13 above)
Comment (ronen-fr) [reply src/crimson/admin/admin_socket.cc]: will do.

**Comment by athanatos** [comment]
Ah, I think I see the confusion.  Classic osd's CephContext is a process singleton (see src/global/global_context.h).  We're explicitly not doing that -- crimson::osd::ShardServices directly contains a CephContext.  If/when we move to multiple reactors, each one will get its own instance of crimson::osd::ShardServices, which is why the other machinery in ShardServices assumes thread local access.

This admin socket implementation is a little different as you do need a process singleton thread for dealing with the admin socket itself, so I'd suggest you create that on the main osd object and have it pass messages to a per-reactor object located within each CephContext::shard_services.

Let me look into this a bit more, perhaps we can do it without much refactoring.

File: src/common/ceph_context.h (hunk 12 above)
Comment (athanatos) [reply src/common/ceph_context.h]: Sorry, I just noticed that you did this for uniformity with the existing interface, that's reasonable.

File: src/crimson/admin/admin_socket.cc [hunk 16]
```diff
@@ -0,0 +1,896 @@
...
+seastar::future<> AdminSocket::init_async(const std::string& path)
+{
+  internal_hooks().get(); // we are executing in an async thread, thus OK to wait()
+
+  logger().debug("{}: asok socket path={}", __func__, path);
+  auto sock_path = seastar::socket_address{seastar::unix_domain_addr{path}};
+
+  return seastar::do_with(seastar::engine().listen(sock_path), [this](seastar::server_socket& lstn) {
+
+    return seastar::do_until([this](){ return do_die; }, [&lstn,this]() {
```
Comment (athanatos) [inline src/crimson/admin/admin_socket.cc]: This all looks like fully async IO?  Do we actually need seastar::async at all?

File: src/crimson/admin/admin_socket.cc (hunk 15 above)
Comment (athanatos) [reply src/crimson/admin/admin_socket.cc]: No, I guess the intention is to ensure that two concurrent admin socket commands don't hit the same hook at the same time?

File: src/crimson/admin/admin_socket.h [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+    Note: uses an async seastar::thread to wait for servers_tbl_rwlock. The returned
+    future is guaranteed to be available.
+   */
+  seastar::future<AdminSocket::GateAndHook> locate_subcmd(std::string match);
+
+public:
+  /*!
+    iterator support
+   */
+  AdminHooksIter begin();
```
Comment (athanatos) [inline src/crimson/admin/admin_socket.h]: I think I'm missing something, where's the user for this?

Comment (ronen-fr) [reply src/crimson/admin/admin_socket.h]: Every range-based for loop, e.g. in HelpHook::exec_command():
  for (const auto& hk_info : *m_as) {..}

File: src/crimson/admin/admin_socket.cc (hunk 16 above)
Comment (ronen-fr) [reply src/crimson/admin/admin_socket.cc]: Probably no. But it:
-  probably helpful when shutting the server down before OSD termination (I'll code this shortly and see),
-  seems cleaner, as this is indeed a separate role

File: src/crimson/admin/admin_socket.cc (hunk 14 above)
Comment (ronen-fr) [reply src/crimson/admin/admin_socket.cc]: will fix

File: src/crimson/admin/admin_socket.cc (hunk 15 above)
Comment (ronen-fr) [reply src/crimson/admin/admin_socket.cc]: To answer the questions regarding the locking decisions:
execute_line() uses parse_cmd() to analyze the incoming message and locate the command in one of the servers' blocks. Once the relevant registered API is located, the general RW-lock on the servers table is released. The only thing that keeps the server, the one that actually handles that registered API, from disappearing from under our feet is the gate we've entered.
Now - the code can be made much simpler if we simply keep the general RW lock at all times while executing an asok command. I can then remove all gates-related handling. I am not sure that that's a bad design decision. It's just that when I wrote the code - the gates came first, as I was trying to avoid the non-granular RW lock.

File: src/crimson/admin/admin_socket.cc (hunk 16 above)
Comment (athanatos) [reply src/crimson/admin/admin_socket.cc]: I'm not sure if this is still relevant, but we probably want to completely avoid seastar::thread unless absolutely necessary.  The performance implications don't actually matter here, but in general seastar::thread is less performant and we don't want to be designing things with that in mind.

**Comment by ronen-fr** [comment]
Replaced by #32174
